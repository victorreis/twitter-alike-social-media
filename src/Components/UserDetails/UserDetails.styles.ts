import styled from 'styled-components';

import { UserDetailsStyleProps } from './UserDetails.types';

export const UserDetailsContainer = styled.div<UserDetailsStyleProps>`
  display: flex;
  flex-direction: column;
  flex: 1;
  padding: 16px;
  row-gap: 1rem;

  border-bottom-width: 1px;
  border-bottom-style: solid;
  border-bottom-color: ${({ theme }) => theme.colors.background.default.normal};
`;

export const UserDetailsHeader = styled.div<UserDetailsStyleProps>`
  display: flex;
  column-gap: 1rem;
`;

export const UserDetailsHeaderContent = styled.div<UserDetailsStyleProps>`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  width: 100%;
  row-gap: 0.5rem;
`;

export const UserDetailsContent = styled.div<UserDetailsStyleProps>`
  display: flex;
  flex-direction: column;
  row-gap: 0.2rem;
`;
