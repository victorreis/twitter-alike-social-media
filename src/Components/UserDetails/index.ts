export * from './UserDetails.types';
export * from './UserDetails';
