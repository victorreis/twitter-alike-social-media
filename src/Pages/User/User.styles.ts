import styled from 'styled-components';

export const UserContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export const UserTabBar = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: flex-start;

  box-sizing: border-box;
  border-bottom: 1px solid
    ${({ theme }) => theme.colors.background.default.light};
`;
