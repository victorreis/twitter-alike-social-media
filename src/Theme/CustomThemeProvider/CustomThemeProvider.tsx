/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
