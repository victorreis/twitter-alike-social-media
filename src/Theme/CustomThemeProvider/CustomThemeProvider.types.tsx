import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
