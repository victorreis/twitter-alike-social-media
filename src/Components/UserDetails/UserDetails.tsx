import { useState } from 'react';

import dayjs from 'dayjs';

import { EXTENDED_DATE_FORMAT } from '../../Config/Constants';
import { TestProps } from '../../Config/Tests/Test.types';
import { FeedbackColor } from '../../Theme/Types/Colors.types';
import { Avatar } from '../Avatar';
import { Button } from '../Button';
import { Typography } from '../Typography';
import {
  UserDetailsContainer,
  UserDetailsContent,
  UserDetailsHeader,
  UserDetailsHeaderContent,
} from './UserDetails.styles';
import { UserDetailsProps, DefaultUserDetailsProps } from './UserDetails.types';

export const userDetailsDefaults: Required<DefaultUserDetailsProps> &
  Required<TestProps> = {
  testID: 'UserDetails',
};

export const UserDetails: React.FC<UserDetailsProps> = (props): JSX.Element => {
  const {
    testID = userDetailsDefaults.testID,
    isFollowing: userIsFollowing,
    onFollow,
    name,
    nickname,
    thumbnailUrl,
    createdAt,
    following,
    followers,
    numberOfPosts,
    style,
    ...others
  } = props;
  const [isFollowing, setIsFollowing] = useState<boolean>(
    Boolean(userIsFollowing)
  );

  const formattedDate = dayjs(createdAt).format(EXTENDED_DATE_FORMAT);
  const buttonText = isFollowing ? 'FOLLOWING' : 'FOLLOW';
  const buttonHoverText = isFollowing ? 'UNFOLLOW' : '';
  const buttonHoverFeedbackColor: FeedbackColor = isFollowing
    ? 'error'
    : 'info';

  const handleFollowClick = () => {
    setIsFollowing((prevState) => {
      if (onFollow) {
        onFollow(!prevState);
      }
      return !prevState;
    });
  };

  return (
    <UserDetailsContainer data-testid={testID} style={style} {...others}>
      <UserDetailsHeader>
        <Avatar name={name} size="LG" thumbnailUrl={thumbnailUrl} />

        <UserDetailsHeaderContent>
          {userIsFollowing !== undefined && (
            <Button
              hoverFeedbackColor={buttonHoverFeedbackColor}
              hoverText={buttonHoverText}
              onClick={handleFollowClick}
            >
              {buttonText}
            </Button>
          )}
          <Typography variant="body2">{numberOfPosts} Posts</Typography>
        </UserDetailsHeaderContent>
      </UserDetailsHeader>

      <UserDetailsContent>
        <Typography>{name}</Typography>
        <Typography variant="body2">@{nickname}</Typography>
        <Typography variant="body2">Joined {formattedDate}</Typography>
        <br />
        <Typography>{following} following</Typography>
        <Typography>{followers} followers</Typography>
      </UserDetailsContent>
    </UserDetailsContainer>
  );
};
