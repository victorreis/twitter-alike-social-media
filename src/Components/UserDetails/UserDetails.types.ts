import { TestProps } from '../../Config/Tests/Test.types';
import { UserType } from '../../Models/User.types';

export interface RequiredUserDetailsProps {
  /**
   * Sets the component styles.
   */
  numberOfPosts: number;
}

export interface DefaultUserDetailsProps {}

export interface OptionalUserDetailsProps {
  /**
   * Defines when the logged user is following or not the loaded user.
   */
  isFollowing?: boolean;

  /**
   * Callback function that is called when the logged user follows the loaded user.
   */
  onFollow?: () => void;

  /**
   * Callback function that is called when the logged user unfollows the loaded user.
   */
  onUnfollow?: () => void;
}

export type UserDetailsProps = RequiredUserDetailsProps &
  DefaultUserDetailsProps &
  OptionalUserDetailsProps &
  TestProps &
  UserType &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type UserDetailsStyleProps = Required<DefaultUserDetailsProps>;
