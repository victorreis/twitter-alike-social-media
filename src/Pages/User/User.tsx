import { useState, useEffect } from 'react';

import { PageContainer } from '../../App.styles';
import { PostCreator } from '../../Components/PostCreator';
import { Tab } from '../../Components/Tab';
import { UserDetails } from '../../Components/UserDetails';
import { TestProps } from '../../Config/Tests/Test.types';
import { useRenderPosts } from '../../Hooks/RenderPosts';
import { useShowUserPage } from '../../Hooks/ShowUserPage';
import { UserType } from '../../Models/User.types';
import { PostTypes } from '../../Services/LocalStorageInitializer';
import { postRetrieverService } from '../../Services/PostRetriever';
import { userRetrieverService } from '../../Services/UserRetriever';
import { UserTabBar } from './User.styles';

export const userDefaults: Required<TestProps> = {
  testID: 'User',
};

export const User: React.FC = (): JSX.Element | null => {
  const [user, setUser] = useState<UserType>();
  const [posts, setPosts] = useState<PostTypes[]>([]);
  const { renderPosts } = useRenderPosts({ posts });

  const { urlNickname, handleCloseUserPage } = useShowUserPage({
    nickname: user?.nickname,
  });

  useEffect(() => {
    const loadedUser = userRetrieverService.getByNickname(String(urlNickname));
    if (!loadedUser) {
      handleCloseUserPage();
      return;
    }
    setUser(() => loadedUser);

    const loadedPosts = postRetrieverService.getAllCreatedByUser(loadedUser.id);
    setPosts(() => loadedPosts);
  }, [handleCloseUserPage, urlNickname]);

  if (!user) {
    return null;
  }

  const { name, thumbnailUrl } = user;

  return (
    <PageContainer data-testid={userDefaults.testID}>
      <UserDetails {...user} />

      <PostCreator
        name={name}
        nickname={String(urlNickname)}
        onChange={() => {}}
        thumbnailUrl={thumbnailUrl}
      />

      <UserTabBar>
        <Tab active text="POSTS" />
      </UserTabBar>

      {renderPosts()}
    </PageContainer>
  );
};
