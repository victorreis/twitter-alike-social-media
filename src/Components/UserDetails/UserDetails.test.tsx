import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { UserType } from '../../User.types';
import { UserDetails, userDetailsDefaults } from './UserDetails';
import { UserDetailsProps } from './UserDetails.types';

describe('userDetails component tests', () => {
  const user: UserType = {
    id: 'user123',
    name: 'Victor Reis',
    nickname: 'victor.reis',
    url: '...',
    thumbnailUrl: 'https://i.pravatar.cc/300',
    createdAt: new Date(),
    following: 333,
    followers: 777,
  };
  const numberOfPosts = 999;

  const requiredProps: UserDetailsProps = {
    ...user,
    numberOfPosts,
  };

  const setup = (props?: UserDetailsProps) => {
    const renderRTR = () =>
      renderRTRCreator<UserDetailsProps>(UserDetails, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<UserDetailsProps>(UserDetails, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(userDetailsDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(user.name);

      expect(element).toBeInTheDocument();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(userDetailsDefaults.testID);

      expect(container).toHaveStyle({
        borderBottomColor: themes.default.colors.background.default.normal,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
