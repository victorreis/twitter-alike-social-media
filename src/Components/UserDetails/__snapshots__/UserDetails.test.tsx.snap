// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userDetails component tests snapshot tests should render correctly 1`] = `
<div
  className="UserDetailsContainer-twitter-alike-social-media__sc-1pkktdh-0 ewYhwp"
  data-testid="UserDetails"
  id="user123"
>
  <div
    className="UserDetailsHeader-twitter-alike-social-media__sc-1pkktdh-1 jPbomt"
  >
    <div
      className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 jOfrBo"
      data-testid="Avatar"
      size="LG"
    >
      <img
        alt="Victor Reis's avatar photo."
        className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 lbyZtm"
        data-testid="Avatar_image"
        onError={[Function]}
        size="LG"
        src="https://i.pravatar.cc/300"
        title="Victor Reis's avatar photo."
      />
    </div>
    <div
      className="UserDetailsHeaderContent-twitter-alike-social-media__sc-1pkktdh-2 cVyYAQ"
    >
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV"
        data-testid="Typography"
      >
        999
         Posts
      </span>
    </div>
  </div>
  <div
    className="UserDetailsContent-twitter-alike-social-media__sc-1pkktdh-3 COovM"
  >
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
      data-testid="Typography"
    >
      Victor Reis
    </span>
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV"
      data-testid="Typography"
    >
      @
      victor.reis
    </span>
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV"
      data-testid="Typography"
    >
      Joined 
      January 1, 2011
    </span>
    <br />
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
      data-testid="Typography"
    >
      333
       following
    </span>
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
      data-testid="Typography"
    >
      777
       followers
    </span>
  </div>
</div>
`;
